### ISSUE TITLE
Broken default C++ version when enabling HIP backend

### ISSUE THREAD

**Issue by etiennemlb** [kokkos/kokkos#5921]
TLDR; The issue emanate from the following:
```
$ cat main.cc 
#include <cstdio>
int main() { std::printf("%ld\n", __cplusplus); }
$ CC main.cc && ./a.out 
201402
$ CC -xhip main.cc && ./a.out 
201103
```
A solution is to force CMake to produce a `-std=c++XX` using, for instance `-DCMAKE_CXX_STANDARD=14`.

The Kokkos documentation specifies `Kokkos_CXX_STANDARD: STRING Default: 14, The C++ standard for Kokkos to use` (note the default).

Assuming the following reproducer based on this CMake script:
```
cmake_minimum_required(VERSION 3.16)
project("FooProj" LANGUAGES CXX)
set(Kokkos_ENABLE_HIP ON CACHE BOOL "")
set(Kokkos_ARCH_VEGA90A ON CACHE BOOL "")
add_subdirectory(kokkos)
add_library(mylib configuration.cc)
target_link_libraries(mylib Kokkos::kokkos)
```
the following C++ code:
```
#include <cstdio>
int main() { std::printf("%ld\n", __cplusplus); }
```
and the CMake configuration and compilation line:
```
cmake -DCMAKE_CXX_COMPILER=CC -G Ninja ..  && ninja
```
We get a compilation error due to the lack of explicit C++ standard specification (`-std=c++XX`). That is, we try to compile C++14 and up code, using a compiler setup for C++11 (bunch of `std::conditional_t<...>` vs `typename std::conditional<...>::type` kind of error).

I show the issue under different compilation commands. For each command, an excerpt of the CMake configuration and Ninja build script is given:

For `cmake -DCMAKE_CXX_COMPILER=CC                          -G Ninja ..  && ninja` (**FAILS** to compile):
CMake:
```
-- Compiler Version: 15.0.0
-- Using -std=gnu++14 for C++14 extensions as feature
```
Ninja:
```
FLAGS = -fno-gpu-rdc -x hip --rocm-path=/opt/rocm-5.2.3 --offload-arch=gfx90a
```

For `cmake -DCMAKE_CXX_COMPILER=CC -DKokkos_CXX_STANDARD=14 -G Ninja ..  && ninja` (**FAILS** to compile):
CMake:
```
-- Compiler Version: 15.0.0
-- Using -std=gnu++14 for C++14 extensions as feature
```
Ninja:
```
FLAGS = -fno-gpu-rdc -x hip --rocm-path=/opt/rocm-5.2.3 --offload-arch=gfx90a
```

For `cmake -DCMAKE_CXX_COMPILER=CC -DKokkos_CXX_STANDARD=17 -G Ninja ..  && ninja`  (**DOES** compile):
CMake:
```
-- Compiler Version: 15.0.0
-- Using -std=gnu++17 for C++17 extensions as feature
```
Ninja:
```
FLAGS = -fno-gpu-rdc -x hip --rocm-path=/opt/rocm-5.2.3 --offload-arch=gfx90a -std=gnu++17
```

For `cmake -DCMAKE_CXX_COMPILER=CC -DCMAKE_CXX_STANDARD=14  -G Ninja ..  && ninja` (**DOES** compile):
CMake:
```
-- Compiler Version: 15.0.0
-- Using -std=gnu++14 for C++14 extensions as feature
```
Ninja:
```
FLAGS = -fno-gpu-rdc -x hip --rocm-path=/opt/rocm-5.2.3 --offload-arch=gfx90a -std=gnu++14
```

For `cmake -DCMAKE_CXX_COMPILER=CC -DCMAKE_CXX_STANDARD=17  -G Ninja ..  && ninja`  (**DOES** compile):
CMake:
```
-- Compiler Version: 15.0.0
-- Using -std=gnu++17 for C++17 extensions as feature
```
Ninja:
```
FLAGS = -fno-gpu-rdc -x hip --rocm-path=/opt/rocm-5.2.3 --offload-arch=gfx90a -std=gnu++17
```

Note that the `-- Using XXX` line in the CMake configuration output is always correct, but it does not translate to the flag being used !

Additional information:

CMake version 3.20.2.
Kokkos version (3.7.01).
ROCm, 5.2.3.

Modules:
```
Currently Loaded Modules:
  1) craype-x86-trento    3) craype-network-ofi       5) cce/15.0.0      7) cray-dsmml/0.2.2    9) cray-libsci/22.11.1.2  11) amd-mixed/5.2.3
  2) libfabric/1.15.2.0   4) perftools-base/22.09.0   6) craype/2.7.19   8) cray-mpich/8.1.21  10) PrgEnv-cray/8.3.3      12) craype-accel-amd-gfx90a
```
You could change `amd-mixed` for a rocm.

**Comment by Rombur** [comment]
Are you using the Cray compiler to compile hip code because that's not supported. I know some people can compile their code with it but as far as I know, the Cray compiler does not support everything we require.

**Comment by etiennemlb** [comment]
That is something I did not know, the [Known issue section](https://kokkos.github.io/kokkos-core-wiki/building.html#known-issues-a-name-knownissues-a) refers to Cray as if it was supported. and the [requirement](https://kokkos.github.io/kokkos-core-wiki/requirements.html) explicitly says that Clang is supported. Clang which to TBH, the HPE-Cray is very very close except for some backend stuff (though I would appreciate remarks regarding the discrepancy when cce and native clang or amdclang).

**Comment by etiennemlb** [comment]
I will try with AMDclang (prgenv-amd) and I believe a similar issue will show up.

**Comment by etiennemlb** [comment]
@rombur Same issue reproduced with `PrgEnv-amd`.

**Comment by etiennemlb** [comment]
The problem with is `CMAKE_CXX_STANDARD` is that its just wayy to pervasive. I do not want Kokkos's implementation details to leak everywhere in a cmake project.

**Comment by etiennemlb** [comment]
Ideally I would use something like:
```
set_target_properties("${target_name}" PROPERTIES CXX_STANDARD 17
                                                  CXX_STANDARD_REQUIRED ON
                                                  CXX_EXTENSIONS OFF)
```
Toi restrict to a given target, the C++ version I want to use

**Comment by etiennemlb** [comment]
@Rombur regarding the OLCF bug report, do you have more details on what was broken ? Appart for this build issue, it seems to work well.

**Comment by Rombur** [comment]
ICE when compiling the testsuite. I know other people had issues when compiling complex codes.

**Comment by etiennemlb** [comment]
@dalg24 I use Kokkos 3.7.01. regarding, CMAKE_CXX_STANDARD, I agree that for simple codes, it works, but ideally, I would like to avoid that (bib big hammer of a variable), and Kokkos should, by default, select C++14 and compile correctly, which it does not  seem to do on my machine.

**Comment by dalg24** [comment]
If you don't specify any, Kokkos will set it to the minimum required.  It does not propagate to parent project.

**Comment by PhilMiller** [comment]
Compiling C++ code with a mix of standards is generally not a great idea, because there can be myriad under- and un-documented ABI compatibility issues. This is more likely to bite when dealing with the very experimental tool chains of new accelerator hardware, mixed with Cray's attempts at 'owning' the compilation process.

Pick a C++ standard version for your whole build, and set it with CMAKE_CXX_STANDARD.

Target properties for C++ standard version can set an acceptable range, but they can't ensure compatibility.

**Comment by etiennemlb** [comment]
@PhilMiller  I'm quite aware of the ABI issues with C++ (though with *LLVMisation* and OS wanting C++ dynlib, this tends to settle). Say I have multiple libraries in a given project, built with the same compiler but multiple standards and wrapped in a C89 API. I still do not want a global C++ version.

@dalg24 This is the issue here, the default provided by Kokkos does not seem to work. The minimum required Kokkos expects to have (C+14) is not the one enabled by default when the `-x hip` flag is set.

**Comment by ajpowelsnl** [comment]
@Rombur -- Are there additional `TODO` for this issue, beyond what you've done in #5945 ?

**Comment by Rombur** [comment]
No
